### PR TITLE
wgengine/netstack: use netip.Prefix as map keys

### PR DESCRIPTION
**Pull request by maisem**

### CODE REVIEW

**Review by raggi — COMMENTED** [review]
I've had to debug this code in two different PRs and each time it's needed lots of printfs / a debugger. I'd be much more comfortable landing this if we had tests.

**Comment by maisem** [comment]
> I've had to debug this code in two different PRs and each time it's needed lots of printfs / a debugger. I'd be much more comfortable landing this if we had tests.

Fair, but I recommend seeing this commit by commit as the change is really just a map key change.

File: wgengine/netstack/netstack.go [hunk 1]
```diff
@@ -343,74 +343,72 @@ func (ns *Impl) UpdateNetstackIPs(nm *netmap.NetworkMap) {
...
+	for p := range oldPfx {
+		if !newPfx[p] {
+			pfxToRemove[p] = true
 		}
 	}
 	ns.mu.Lock()
 	for ip := range ns.connsOpenBySubnetIP {
-		ipp := tcpip.AddrFromSlice(ip.AsSlice()).WithPrefix()
-		delete(ipsToBeRemoved, ipp)
+		// TODO(maisem): this looks like a bug, remove or document.
```
Comment (tendstofortytwo) [inline wgengine/netstack/netstack.go:378]: what looks like a bug here?

Comment (maisem) [reply wgengine/netstack/netstack.go:378]: That we are not cleaning up some addresses. I mostly just want it documented

File: wgengine/netstack/netstack.go [hunk 2]
```diff
@@ -362,48 +362,46 @@ func (ns *Impl) UpdateNetstackIPs(nm *netmap.NetworkMap) {
...
+		err := ns.ipstack.RemoveAddress(nicID, tcpip.AddrFromSlice(ipp.Addr().AsSlice()))
 		if err != nil {
 			ns.logf("netstack: could not deregister IP %s: %v", ipp, err)
 		} else {
 			ns.logf("[v2] netstack: deregistered IP %s", ipp)
 		}
 	}
 	for ipp := range ipsToBeAdded {
+		if !ipp.IsValid() {
+			ns.logf("netstack: skipping invalid IP %s", ipp)
```
Comment (bradfitz) [inline wgengine/netstack/netstack.go]: if it's not valid, what different forms might this stringify as that you're using %s?

File: wgengine/netstack/netstack.go [hunk 3]
```diff
@@ -362,48 +362,46 @@ func (ns *Impl) UpdateNetstackIPs(nm *netmap.NetworkMap) {
...
 			pa.Protocol = ipv6.ProtocolNumber
-		case 4:
+		} else {
 			pa.Protocol = ipv4.ProtocolNumber
-		default:
-			ns.logf("[unexpected] netstack: could not register IP %s without protocol: unknown IP length (%v)", ipp, ipp.Address.Len())
-			continue
 		}
-		var err tcpip.Error
-		err = ns.ipstack.AddProtocolAddress(nicID, pa, stack.AddressProperties{
```
Comment (bradfitz) [inline wgengine/netstack/netstack.go:406]: this was explicitly written like this to make it very obvious `err` isn't an `error`.

Comment (maisem) [reply wgengine/netstack/netstack.go:406]: the `err` threw me off, will re-add as `var tcpErr tcpip.Error`

File: wgengine/netstack/netstack.go (hunk 1 above)
Comment (bradfitz) [reply wgengine/netstack/netstack.go:378]: What's "it"? Naman is saying that "this" isn't clear.

**Review by bradfitz — APPROVED** [review]
LGTM after addressing those things

File: wgengine/netstack/netstack.go (hunk 1 above)
Comment (maisem) [reply wgengine/netstack/netstack.go:378]: >That we are not cleaning up some addresses.

It seems like we are leaking them, or removing them accidentally. e.g. if we are advertising `1.2.3.4/32`, and there is an active connection for `1.2.3.4` we won't remove it even if we stop advertising `1.2.3.4`. On the other hand, we might end up removing it when the connection counts drops to 0?

File: wgengine/netstack/netstack.go (hunk 2 above)
Comment (maisem) [reply wgengine/netstack/netstack.go]: change to log explicitly as `p.Addr()/p.Bits()`

**Pull request merged**